File: examples/csharp/Prj/CloudEvents/PlayerCloudEventsProjector.cs
using Google.Protobuf.WellKnownTypes;
using Angzarr.Client;
using Angzarr.Proto.Angzarr;
using Angzarr.Proto.Examples;

namespace Angzarr.Examples.PrjCloudEvents;

/// <summary>
/// CloudEvents projector - publishes player events as CloudEvents.
///
/// This projector transforms internal domain events into CloudEvents 1.0 format
/// for external consumption via HTTP webhooks or Kafka.
/// </summary>

// docs:start:cloudevents_oo
public class PlayerCloudEventsProjector : CloudEventsProjector
{
    public PlayerCloudEventsProjector()
        : base("prj-player-cloudevents", "player") { }

    [Publishes(typeof(PlayerRegistered))]
    public CloudEvent? OnPlayerRegistered(PlayerRegistered @event)
    {
        // Filter sensitive fields, return public version
        var publicEvent = new PublicPlayerRegistered
        {
            DisplayName = @event.DisplayName,
            PlayerType = @event.PlayerType
        };
        return new CloudEvent
        {
            Type = "com.poker.player.registered",
            Data = Any.Pack(publicEvent)
        };
    }

    [Publishes(typeof(FundsDeposited))]
    public CloudEvent? OnFundsDeposited(FundsDeposited @event)
    {
        var publicEvent = new PublicFundsDeposited
        {
            Amount = @event.Amount
        };
        return new CloudEvent
        {
            Type = "com.poker.player.deposited",
            Data = Any.Pack(publicEvent),
            Extensions = { ["priority"] = "normal" }
        };
    }
}
// docs:end:cloudevents_oo

// docs:start:cloudevents_router
public static class PlayerCloudEventsHandlers
{
    public static CloudEvent? HandlePlayerRegistered(PlayerRegistered @event)
    {
        var publicEvent = new PublicPlayerRegistered
        {
            DisplayName = @event.DisplayName,
            PlayerType = @event.PlayerType
        };
        return new CloudEvent
        {
            Type = "com.poker.player.registered",
            Data = Any.Pack(publicEvent)
        };
    }

    public static CloudEvent? HandleFundsDeposited(FundsDeposited @event)
    {
        var publicEvent = new PublicFundsDeposited
        {
            Amount = @event.Amount
        };
        return new CloudEvent
        {
            Type = "com.poker.player.deposited",
            Data = Any.Pack(publicEvent),
            Extensions = { ["priority"] = "normal" }
        };
    }

    public static CloudEventsRouter BuildRouter() =>
        new CloudEventsRouter("prj-player-cloudevents", "player")
            .On<PlayerRegistered>(HandlePlayerRegistered)
            .On<FundsDeposited>(HandleFundsDeposited);
}
// docs:end:cloudevents_router

File: examples/java/prj-cloudevents/src/main/java/dev/angzarr/examples/prjcloudevents/PlayerCloudEventsProjector.java
package dev.angzarr.examples.prjcloudevents;

import com.google.protobuf.Any;
import dev.angzarr.client.CloudEventsProjector;
import dev.angzarr.client.CloudEventsRouter;
import dev.angzarr.proto.angzarr.CloudEvent;
import dev.angzarr.proto.examples.FundsDeposited;
import dev.angzarr.proto.examples.PlayerRegistered;
import dev.angzarr.proto.examples.PublicFundsDeposited;
import dev.angzarr.proto.examples.PublicPlayerRegistered;

/**
 * CloudEvents projector - publishes player events as CloudEvents.
 *
 * This projector transforms internal domain events into CloudEvents 1.0 format
 * for external consumption via HTTP webhooks or Kafka.
 */

// docs:start:cloudevents_oo
public class PlayerCloudEventsProjector extends CloudEventsProjector {

    public PlayerCloudEventsProjector() {
        super("prj-player-cloudevents", "player");
    }

    @Publishes("PlayerRegistered")
    public CloudEvent onPlayerRegistered(PlayerRegistered event) {
        // Filter sensitive fields, return public version
        var publicEvent = PublicPlayerRegistered.newBuilder()
            .setDisplayName(event.getDisplayName())
            .setPlayerType(event.getPlayerType())
            .build();
        return CloudEvent.newBuilder()
            .setType("com.poker.player.registered")
            .setData(Any.pack(publicEvent))
            .build();
    }

    @Publishes("FundsDeposited")
    public CloudEvent onFundsDeposited(FundsDeposited event) {
        var publicEvent = PublicFundsDeposited.newBuilder()
            .setAmount(event.getAmount())
            .build();
        return CloudEvent.newBuilder()
            .setType("com.poker.player.deposited")
            .setData(Any.pack(publicEvent))
            .putExtensions("priority", "normal")
            .build();
    }
}
// docs:end:cloudevents_oo
